nexus-api/main.go

```diff
@@ -1,20 +1,118 @@
 package main
 
 import (
+	"encoding/json"
 	"fmt"
 	"net/http"
 	"nexus-api/service"
+
+	"github.com/gorilla/mux"
+	"golang.org/x/crypto/bcrypt"
 )
 
 func main() {
+	fmt.Println("api server starting")
+
+	router := mux.NewRouter()
+
 	// attach router to default http server mux
-	http.Handle("/", service.CorsMiddleware(HelloServer))
+	router.HandleFunc("/login", service.CorsMiddleware(LoginHandler))
+	router.HandleFunc("/hello", service.CorsMiddleware(HelloServer))
+
+	// attach router to default http server mux
+	http.Handle("/", router)
 
 	http.ListenAndServe(":8080", nil)
+
 }
 
 func HelloServer(w http.ResponseWriter, r *http.Request) {
 	name := r.URL.Path[1:]
 	fmt.Printf("api called with %s \n", name)
 	fmt.Fprintf(w, "Hello, %s!", name)
 }
+
+type LoginRequest struct {
+	Password string `json:"password"`
+	Username string `json:"username"`
+}
+
+type LoginResponse struct {
+	RedirectURL string `json:"redirect_url"`
+	Password    string `json:"password"`
+	Username    string `json:"username"`
+	Hash        string `json:"hash"`
+	Match       bool   `json:"match"`
+}
+
+var LoginInfo = map[string]string{
+	"abdul":     "passwordHash",
+	"username2": "passwordHash",
+}
+
+func HashPassword(password string) (string, error) {
+	bytes, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)
+	return string(bytes), err
+}
+
+func CheckPasswordHash(password, hash string) bool {
+	err := bcrypt.CompareHashAndPassword([]byte(hash), []byte(password))
+	return err == nil
+}
+
+func LoginHandler(w http.ResponseWriter, r *http.Request) {
+	var request LoginRequest
+	err := json.NewDecoder(r.Body).Decode(&request)
+
+	if err != nil {
+		fmt.Printf("error %s parsing %+v", err, request)
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusBadRequest)
+		json.NewEncoder(w).Encode(struct{}{})
+		return
+	}
+
+	fmt.Printf("login username %s, password %s\n", request.Username, request.Password)
+
+	passwordHashForUser, exists := LoginInfo[request.Username]
+	if !exists {
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusUnauthorized)
+		json.NewEncoder(w).Encode(struct{}{})
+		return
+	}
+
+	match := CheckPasswordHash(request.Password, passwordHashForUser)
+
+	response := LoginResponse{
+		RedirectURL: "/",
+		Password:    request.Password,
+		Username:    request.Username,
+		Hash:        passwordHashForUser,
+		Match:       match,
+	}
+
+	// username doesn't exist in our system
+	if !exists {
```

Comment:
from here to line 102 can be deleted / are duplicated by lines 78 - 83

Reply:
ah I think I see what you were trying to do, here you should check `if !match` vs the exists variable which is tied to whether the username exists